ledger/simulation/resources.go

```diff
@@ -0,0 +1,607 @@
+	if globalSharing {
+		return a.globalResources.addApp(aid, ep, programVersion)
+	}
+	if !a.localTxnResources[gi].addApp(aid, ep, programVersion) {
+		return false
+	}
+	if a.globalResources.hasApp(aid) {
+		// It's redundant to list a resources in both the global and local tracker, so remove it
+		// from global. The below call to a.globalResources.removeAppSlot() will revert the changes
+		// to a.globalResources.MaxApps and a.globalResources.MaxTotalRefs.
+		delete(a.globalResources.Apps, aid)
+		a.globalResources.MaxApps++
+		a.globalResources.MaxTotalRefs++
+	}
+	// This ensures that the global tracker reduces in size if a resource is assigned locally.
+	if a.globalResources.removeAppSlot() {
+		return true
+	}
+	// Undo the local assignment if global is full.
+	delete(a.localTxnResources[gi].Apps, aid)
+	return false
+}
+
+func (a *groupResourceTracker) hasBox(app basics.AppIndex, name string) bool {
+	// All boxes are global, never consult localTxnResources
+	return a.globalResources.hasBox(app, name)
+}
+
+func (a *groupResourceTracker) addBox(app basics.AppIndex, name string, readSize, additionalReadBudget, bytesPerBoxRef uint64) bool {
+	// All boxes are global, never consult localTxnResources
+	return a.globalResources.addBox(app, name, readSize, additionalReadBudget, bytesPerBoxRef)
+}
+
+func (a *groupResourceTracker) reconcileBoxWriteBudget(used uint64, bytesPerBoxRef uint64) error {
+	if !a.globalResources.addEmptyBoxRefsForWriteBudget(used, uint64(a.startingBoxes)*bytesPerBoxRef, bytesPerBoxRef) {
+		return fmt.Errorf("cannot add extra box refs to satisfy write budget of %d bytes", used)
+	}
+	return nil
+}
+
+func (a *groupResourceTracker) maxPossibleBoxIOBudget(bytesPerBoxRef uint64) uint64 {
+	return basics.MulSaturate(
+		uint64(a.startingBoxes+a.globalResources.maxPossibleUnnamedBoxes()),
+		bytesPerBoxRef,
+	)
+}
+
+func (a *groupResourceTracker) hasHolding(addr basics.Address, aid basics.AssetIndex) bool {
+	// All cross-products are global, never consult localTxnResources
+	return a.globalResources.hasHolding(addr, aid)
+}
+
+func (a *groupResourceTracker) addHolding(addr basics.Address, aid basics.AssetIndex) bool {
+	// All cross-products are global, never consult localTxnResources
+	return a.globalResources.addHolding(addr, aid)
+}
+
+func (a *groupResourceTracker) hasLocal(addr basics.Address, aid basics.AppIndex, ep *logic.EvalParams) bool {
+	// All cross-products are global, never consult localTxnResources
+	return a.globalResources.hasLocal(addr, aid, ep)
+}
+
+func (a *groupResourceTracker) addLocal(addr basics.Address, aid basics.AppIndex) bool {
+	// All cross-products are global, never consult localTxnResources
+	return a.globalResources.addLocal(addr, aid)
+}
+
+type resourcePolicy struct {
+	tracker                     groupResourceTracker
+	ep                          *logic.EvalParams
+	initialBoxSurplusReadBudget *uint64
```

Comment:
nit: it is a bit confusing that `initialBoxSurplusReadBudget`, as implement-wise, it occurred 4 times, 2 in `resources.go` and 2 in `tracer.go`.  It only get assigned from `cx.SurplusReadBudget` in `tracer.go`, and used in `addBox` in `resources.go`, which is a method in `*resourcePolicy`.

so my question is, why we not let this be an instance rather than a ptr?  This is related with line 421 in tracer.go.

Reply:
The reason it's a pointer is so that we can tell if it's been set initially or not. It's realistic that this value could be 0, so I didn't want to rely on that as an indicator
https://github.com/algorand/go-algorand/blob/128725e99fd91ff1724b96ced44e8d0a6c5e4917/ledger/simulation/tracer.go#L421-L424

Though I realize now that there's no harm in resetting it every time -- `cx.SurplusReadBudget` should stay the same for the entirety of the group evaluation. (I wrote this code before exposing `cx.SurplusReadBudget`, and it used to calculate the value independently, which I didn't want to repeat unnecessarily, but that's no longer an issue now.)

So yes I think this could be changed but I'm not sure it's worth the trouble at this point.

Reply:
Thanks for explanation, I was thinking about the intentions behind the code, and that makes sense to me.